core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -561,25 +601,61 @@ public boolean renameDirectory(String src, String dst) throws IOException {
       return false;
     }
     // Rename each child in the src folder to destination/child
+    // a. Since renames are a copy operation, files are added to a buffer and processed concurrently
+    // b. Pseudo-directories are metadata only operations are not added to the buffer
+    RenameBuffer buffer = new RenameBuffer();
     for (UfsStatus child : children) {
       String childSrcPath = PathUtils.concatPath(src, child.getName());
       String childDstPath = PathUtils.concatPath(dst, child.getName());
-      boolean success;
       if (child.isDirectory()) {
         // Recursive call
-        success = renameDirectory(childSrcPath, childDstPath);
+        if (!renameDirectory(childSrcPath, childDstPath)) {
+          LOG.error("Failed to rename path {} to {}, aborting rename.", childSrcPath, childDstPath);
+          return false;
+        }
       } else {
-        success = renameFile(childSrcPath, childDstPath);
-      }
-      if (!success) {
-        LOG.error("Failed to rename path {} to {}, aborting rename.", childSrcPath, childDstPath);
-        return false;
+        buffer.add(new Pair<>(childSrcPath, childDstPath));
       }
     }
+    // Get result of parallel file renames
+    int filesRenamed = buffer.getResult().size();
+    if (filesRenamed != buffer.mEntriesAdded) {
+      LOG.warn("Failed to rename directory, successfully renamed {} files out of {}.",
+          filesRenamed, buffer.mEntriesAdded);
+      return false;
+    }
     // Delete src and everything under src
     return deleteDirectory(src, DeleteOptions.defaults().setRecursive(true));
   }
 
+  /**
+   * File paths added to a {@link RenameBuffer} will be renamed concurrently.
+   */
+  @NotThreadSafe
+  protected class RenameBuffer extends OperationBuffer<Pair<String, String>> {
+    /**
+     * Construct a new {@link RenameBuffer} instance.
+     */
+    public RenameBuffer() {}
+
+    @Override
+    protected int getBatchSize() {
+      return 1;
```

Comment:
why choose the batch size as 1?
The rename buffer targets the big directory with hundreds/thousands of files. When we have thousands of files, by default the executor is a fixed 20 thread thread pool. In this case, it would be better to increase the batch size.

Reply:
the batch determines how operands are grouped before submitting to executor pool. for deletes, each s3a call (bulk delete) operates on 1000 objects by default. for rename, each s3a call (copy) operates on a single object.

the batch size of 1 means that we submit a job to the executor pool w/o batching, so for a directory with 10k files, 10k jobs will be submitted to be processed concurrently by 20 threads.

Reply:
Got it, thanks for explanation!

Reply:
I didn't realize that we cannot do batch copy object. It will be helpful if s3 provides the batch operations.https://aws.amazon.com/about-aws/whats-new/2018/11/s3-batch-operations/

Reply:
yeah, i didn't find a bulk copy api in the java sdk

Reply:
thx Lu. let me look into if we can invoke a batch copy job using the java sdk